server/store/user_code.go

```
@@ -0,0 +1,109 @@
+package store
+
+import (
+	sq "github.com/Masterminds/squirrel"
+	"github.com/oseducation/knowledge-graph/model"
+	"github.com/pkg/errors"
+)
+
+// UserCodeStore is an interface to crud users's codes
+type UserCodeStore interface {
+	Save(userCode *model.UserNodeCode) error
+	Update(new *model.UserNodeCode) error
+	GetCodesForUserNode(userID, nodeID string) ([]*model.UserNodeCode, error)
+	GetCodeForUserNode(userID, nodeID, codeName string) (*model.UserNodeCode, error)
+}
+
+// SQLUserCodeStore is a struct to store users's codes
+type SQLUserCodeStore struct {
+	sqlStore        *SQLStore
+	userCodesSelect sq.SelectBuilder
+}
+
+// NewUserCodeStore creates a new store for user's codes.
+func NewUserCodeStore(db *SQLStore) UserCodeStore {
+	userCodesSelect := db.builder.
+		Select(
+			"uc.user_id",
+			"uc.node_id",
+			"uc.code_name",
+			"uc.code",
+		).
+		From("user_node_codes uc")
+
+	return &SQLUserCodeStore{
+		sqlStore:        db,
+		userCodesSelect: userCodesSelect,
+	}
+}
+
+func (ucs *SQLUserCodeStore) Save(userCode *model.UserNodeCode) error {
+	if err := userCode.IsValid(); err != nil {
+		return err
+	}
+
+	oldCode, err := ucs.GetCodeForUserNode(userCode.UserID, userCode.NodeID, userCode.CodeName)
+	if err != nil && oldCode.Code != userCode.Code {
+		return ucs.Update(userCode)
+	}
```

Should we return an error when the code exists?